instrumentation/kafka-clients/src/main/java/brave/kafka/clients/KafkaTracing.java

```diff
@@ -134,27 +140,44 @@ public Tracing tracing() {
    * one couldn't be extracted.
    */
   public Span nextSpan(ConsumerRecord<?, ?> record) {
-    TraceContextOrSamplingFlags extracted = extractAndClearHeaders(record.headers());
-    Span result = tracing.tracer().nextSpan(extracted);
+    // Eventhough the type is ConsumerRecord, this is not a (remote) consumer span. Only "poll"
+    // events create consumer spans. Since this is a processor span, we use the normal sampler.
+    TraceContextOrSamplingFlags extracted =
+      extractAndClearHeaders(processorExtractor, record.headers(), record.headers());
+    Span result = tracer.nextSpan(extracted);
     if (extracted.context() == null && !result.isNoop()) {
       addTags(record, result);
     }
     return result;
   }
 
-  TraceContextOrSamplingFlags extractAndClearHeaders(Headers headers) {
-    TraceContextOrSamplingFlags extracted = extractor.extract(headers);
-    // clear propagation headers if we were able to extract a span
+  <R> TraceContextOrSamplingFlags extractAndClearHeaders(
+    Extractor<R> extractor, R request, Headers headers
+  ) {
+    TraceContextOrSamplingFlags extracted = extractor.extract(request);
+    // Clear any propagation keys present in the headers
     if (!extracted.equals(TraceContextOrSamplingFlags.EMPTY)) {
       clearHeaders(headers);
     }
     return extracted;
   }
 
-  // BRAVE6: consider a messaging variant of extraction which clears headers as they are read.
-  // this could prevent having to go back and clear them later. Another option is to encourage,
-  // then special-case single header propagation. When there's only 1 propagation key, you don't
-  // need to do a loop!
+  /** Creates a potentially noop remote span representing this request */
+  Span nextMessagingSpan(
+    SamplerFunction<MessagingRequest> sampler,
+    MessagingRequest request,
+    TraceContextOrSamplingFlags extracted
+  ) {
+    Boolean sampled = extracted.sampled();
+    // only recreate the context if the messaging sampler made a decision
+    if (sampled == null && (sampled = sampler.trySample(request)) != null) {
+      extracted = extracted.sampled(sampled.booleanValue());
```

Comment:
sweet! so this is the place where sampling will be overwritten if `samplerFunction` is provided, effectively solving the issue I post a few days ago in gitter :)

Reply:
yep though primary sampling is only at the start of a trace